File: .github/scripts/generate_sitemap.py
"""Gera public/sitemap.xml com todas as páginas + posts do blog."""
import os, json, glob
from datetime import datetime, timezone

BASE_URL = "https://ocodigoaguia.com.br"
posts_dir = "public/blog-posts"
json_files = glob.glob(f"{posts_dir}/post-*.json")
now = datetime.now(timezone.utc).strftime("%Y-%m-%d")

STATIC_PAGES = [
    {"loc": "/", "changefreq": "daily", "priority": "1.0"},
    {"loc": "/blog", "changefreq": "daily", "priority": "0.9"},
    {"loc": "/ebook", "changefreq": "weekly", "priority": "0.8"},
]

urls = []
for page in STATIC_PAGES:
    urls.append(
        f"  <url>\n"
        f"    <loc>{BASE_URL}{page['loc']}</loc>\n"
        f"    <lastmod>{now}</lastmod>\n"
        f"    <changefreq>{page['changefreq']}</changefreq>\n"
        f"    <priority>{page['priority']}</priority>\n"
        f"  </url>"
    )

for jf in sorted(json_files, reverse=True):
    try:
        with open(jf, encoding="utf-8") as f:
            post = json.load(f)
        slug = post.get("slug", os.path.splitext(os.path.basename(jf))[0])
        date = post.get("date", now)
        urls.append(
            f"  <url>\n"
            f"    <loc>{BASE_URL}/blog/{slug}</loc>\n"
            f"    <lastmod>{date}</lastmod>\n"
            f"    <changefreq>weekly</changefreq>\n"
            f"    <priority>0.7</priority>\n"
            f"  </url>"
        )
    except Exception as e:
        print(f"Erro ao processar {jf}: {e}")

sitemap = (
    '<?xml version="1.0" encoding="UTF-8"?>\n'
    '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'
    + "\n".join(urls) +
    "\n</urlset>"
)

os.makedirs("public", exist_ok=True)
with open("public/sitemap.xml", "w", encoding="utf-8") as f:
    f.write(sitemap)

print(f"Sitemap gerado com {len(urls)} URLs em public/sitemap.xml")

File: scripts/generate-sitemap.mjs
/**
 * Generates public/sitemap.xml from src/data/blog.ts
 * Run manually: node scripts/generate-sitemap.mjs
 * Also triggered automatically by GitHub Actions on every push that changes blog.ts
 */
import { readFileSync, writeFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const BASE_URL = "https://ocodigoaguia.com.br";
const TODAY = new Date().toISOString().split("T")[0];

// Parse slugs + dates directly from blog.ts source (no TS compilation needed)
const blogSrc = readFileSync(
  resolve(__dirname, "../src/data/blog.ts"),
  "utf-8"
);

const slugMatches = [...blogSrc.matchAll(/slug:\s*["']([^"']+)["']/g)];
const dateMatches = [...blogSrc.matchAll(/date:\s*["']([\d-]+)["']/g)];

const posts = slugMatches.map((m, i) => ({
  slug: m[1],
  date: dateMatches[i]?.[1] ?? TODAY,
}));

const staticPages = [
  { loc: "/", lastmod: TODAY, changefreq: "weekly", priority: "1.0" },
  { loc: "/blog", lastmod: TODAY, changefreq: "daily", priority: "0.9" },
];

const postPages = posts.map((p) => ({
  loc: `/blog/${p.slug}`,
  lastmod: p.date,
  changefreq: "monthly",
  priority: "0.8",
}));

const allPages = [...staticPages, ...postPages];

const urlEntries = allPages
  .map(
    (p) => `  <url>
    <loc>${BASE_URL}${p.loc}</loc>
    <lastmod>${p.lastmod}</lastmod>
    <changefreq>${p.changefreq}</changefreq>
    <priority>${p.priority}</priority>
  </url>`
  )
  .join("\n");

const xml = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
  http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
${urlEntries}
</urlset>
`;

const outPath = resolve(__dirname, "../public/sitemap.xml");
writeFileSync(outPath, xml, "utf-8");
console.log(`✅ sitemap.xml gerado com ${posts.length} posts → ${outPath}`);
